File: src/web/Utils/IdentityExtensions.cs
using System;
using System.Security.Claims;
using core.fs.Shared.Adapters.Storage;
using core.fs.Shared.Domain.Accounts;

namespace web.Utils
{
    public static class IdentityExtensions
    {
        public const string ID_CLAIM_NAME = "userid";

        public static UserId Identifier(this ClaimsPrincipal p)
        {
            var guid = GetClaimValue(p, ID_CLAIM_NAME);
            if (guid == null)
            {
                throw new Exception($"User is not authenticated. Missing claim: {ID_CLAIM_NAME}");
            }

            return UserId.NewUserId(new Guid(guid));
        }

        public static string Email(this System.Security.Claims.ClaimsPrincipal p) 
            => GetClaimValue(p, ClaimTypes.Email);

        public static string Firstname(this System.Security.Claims.ClaimsPrincipal p) 
            => GetClaimValue(p, ClaimTypes.GivenName);

        public static string Lastname(this System.Security.Claims.ClaimsPrincipal p) 
            => GetClaimValue(p, ClaimTypes.Surname);

        private static string GetClaimValue(ClaimsPrincipal p, string name)
        {
            if (p == null)
            {
                return null;
            }

            var claim = p.FindFirst(name);
            if (claim == null)
            {
                return null;
            }

            return claim.Value;
        }
    }
}

File: src/web.fs/Utils/IdentityExtensions.fs
namespace web.Utils

open System
open System.Security.Claims
open core.fs.Accounts

[<AutoOpen>]
module IdentityExtensions =
    
    [<Literal>]
    let ID_CLAIM_NAME = "userid"
    
    let private getClaimValue (p: ClaimsPrincipal) (name: string) =
        if isNull p then
            None
        else
            let claim = p.FindFirst(name)
            if isNull claim then
                None
            else
                Some claim.Value
    
    type ClaimsPrincipal with
        member this.Identifier() =
            match getClaimValue this ID_CLAIM_NAME with
            | Some guid -> UserId(Guid(guid))
            | None -> raise (Exception($"User is not authenticated. Missing claim: {ID_CLAIM_NAME}"))
        
        member this.Email() =
            match getClaimValue this ClaimTypes.Email with
            | Some email -> email
            | None -> null
        
        member this.Firstname() =
            match getClaimValue this ClaimTypes.GivenName with
            | Some name -> name
            | None -> null
        
        member this.Lastname() =
            match getClaimValue this ClaimTypes.Surname with
            | Some name -> name
            | None -> null
